From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kuan-Chih Wang <kuanchihwang@outlook.com>
Date: Sat, 3 Jan 2026 20:54:04 +0800
Subject: [PATCH] Fix feature detection logic for netCDF-4

As of netCDF-C version 4.9.3, NC_HAS_NC4 is no longer defined.
netCDF-4 is now detected through NC_HAS_HDF5.

The original logic, which relied on NC_VERSION_PATCH, was also
flawed anyway because it did not check NC_VERSION_{MAJOR,MINOR}.

The support status of netCDF-4 is now correctly reflected in
`libpio.settings`.
---
 CMakeLists.txt          | 2 +-
 configure.ac            | 2 +-
 src/clib/pioc_support.c | 7 ++++---
 3 files changed, 6 insertions(+), 5 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 7cb025be..7d360f43 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -321,7 +321,7 @@ SET(STATUS_PNETCDF ${PnetCDF_C_FOUND})
 ###
 CHECK_C_SOURCE_COMPILES("
 #include <netcdf_meta.h>
-#if !NC_HAS_NC4
+#if !NC_HAS_HDF5
       choke me
 #endif
 int main() {return 0;}" HAVE_NETCDF4)
diff --git a/configure.ac b/configure.ac
index d44688da..5207a6a2 100644
--- a/configure.ac
+++ b/configure.ac
@@ -261,7 +261,7 @@ fi
 
 # Do we have netCDF-4?
 AC_COMPILE_IFELSE([AC_LANG_PROGRAM([#include "netcdf_meta.h"],
-[[#if !NC_HAS_NC4
+[[#if !NC_HAS_HDF5
 # error
 #endif]
 ])], [have_netcdf4=yes], [have_netcdf4=no])
diff --git a/src/clib/pioc_support.c b/src/clib/pioc_support.c
index c9139196..93350237 100644
--- a/src/clib/pioc_support.c
+++ b/src/clib/pioc_support.c
@@ -3126,11 +3126,12 @@ iotype_is_valid(int iotype)
         ret++;
 
     /* Some builds include netCDF-4. */
-    /* as of netcdf 4.9.3 NC_HAS_NC4 is no longer defined */
-#if NC_HAS_NC4 || (NC_VERSION_PATCH > 2)
+    /* As of netCDF-C version 4.9.3, NC_HAS_NC4 is no longer defined.
+       netCDF-4 is now detected through NC_HAS_HDF5. */
+#if NC_HAS_HDF5
     if (iotype == PIO_IOTYPE_NETCDF4C || iotype == PIO_IOTYPE_NETCDF4P)
         ret++;
-#endif /* _NETCDF4 */
+#endif /* NC_HAS_HDF5 */
 
     /* Some builds include pnetcdf. */
 #ifdef _PNETCDF
-- 
2.47.3

